lib/Dialect/SV/Transforms/SVExtractTestCode.cpp

```diff
@@ -0,0 +1,224 @@
+// is then instantiated in the original module.
+//
+//===----------------------------------------------------------------------===//
+
+#include "SVPassDetail.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "circt/Dialect/SV/SVVisitors.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+
+#include <set>
+
+using namespace mlir;
+using namespace circt;
+using namespace sv;
+
+//===----------------------------------------------------------------------===//
+// StubExternalModules Helpers
+//===----------------------------------------------------------------------===//
+
+// Compute the dataflow for a set of ops.
+static void dataflowSlice(ArrayRef<Operation *> ops,
+                          SetVector<Operation *> &results) {
+  for (auto op : ops) {
+    getBackwardSlice(op, &results, [](Operation *testOp) -> bool {
+      return !isa<sv::ReadInOutOp>(testOp) && !isa<hw::InstanceOp>(testOp) &&
+             !isa<sv::PAssignOp>(testOp) && !isa<sv::BPAssignOp>(testOp);
+    });
+  }
+}
+
+// Compute the ops defining the blocks a set of ops are in.
+static void blockSlice(ArrayRef<Operation *> ops,
+                       SetVector<Operation *> &blocks) {
+  for (auto op : ops) {
+    while (!isa<hw::HWModuleOp>(op->getParentOp())) {
+      op = op->getParentOp();
+      blocks.insert(op);
+    }
+  }
+}
+
+// Aggressively mark operations to be moved to the new module.  This leaves
+// maximum flexibility for optimization after removal of the nodes from the old
+// module.
+static SetVector<Operation *> computeCloneSet(ArrayRef<Operation *> roots) {
+  SetVector<Operation *> results;
+  // Get Dataflow for roots
+  dataflowSlice(roots, results);
+
+  // Get Blocks
+  SetVector<Operation *> blocks;
+  blockSlice(roots, blocks);
+  blockSlice(results.getArrayRef(), blocks);
+
+  // Make sure dataflow to block args (if conds, etc) is included
+  dataflowSlice(blocks.getArrayRef(), results);
+
+  // include the blocks and roots to clone
+  results.insert(roots.begin(), roots.end());
+  results.insert(blocks.begin(), blocks.end());
+
+  return results;
+}
+
+// Given a set of values, construct a module and bind instance of that module
+// that passes those values through.  Returns the new module and the instance
+// pointing to it.
+static hw::HWModuleOp createModuleForCut(hw::HWModuleOp op,
+                                         SetVector<Value> &inputs,
+                                         BlockAndValueMapping &cutMap,
+                                         StringRef suffix, StringRef path) {
+  OpBuilder b(op->getParentOfType<mlir::ModuleOp>()->getRegion(0));
+
+  // Construct the ports, this is just the input Values
+  SmallVector<hw::ModulePortInfo> ports;
+  for (auto port : llvm::enumerate(inputs))
+    ports.push_back(
+        {b.getStringAttr(""), hw::INPUT, port.value().getType(), port.index()});
```

Comment:
as a future enhancement it might be nice to add some heuristics that grub around the IR to try to find a port name, e.g. if you're referring to a wire, you could use it for the port name etc.